\input{arbeit-vorlage-praeambel.tex} % Importiere die Einstellungen aus der Präambel
% hier beginnt der eigentliche Inhalt

\author{Lydia Buntrock}
\title{master thesis}
\date{Januar 2018}

\hfuzz=\maxdimen \tolerance=10000 \hbadness=10000
% \usepackage[showframe=true]{geometry}


\begin{document}
  % Titelseite
  \begin{titlepage}
    \pagestyle{empty}
  	\begin{center}
      {\Large Freie Universität Berlin} \\
    	\begin{Huge}
      	Fachbereich Mathematik und Informatik \\
      	\vspace{3mm}
    	\end{Huge}
    	\vspace{2cm}
    	\begin{Large}
        \textbf{An analysis of maximum parsimony algorithms to predict parasitism in Eukaryota} \\
        \vspace{3mm}
        using a large multifurcated phylogenetic synthesis tree \\
    	\end{Large}
      \vspace{3cm}
      \textbf{Submitted on:} \\
      3 April 2018 \\
    	\vspace{2cm}
    	Lydia Buntrock \\
      E-Mail: info@irallia.de \\
     	\vspace{3cm}
      \textbf{Supervisor:} \\
      Prof. Dr. Bernhard Y. Renard \\
      \& \\
      Prof. Dr. rer. nat. Emanuel Heitlinger \\      
  	\end{center}
    \clearpage
    \pagenumbering{Roman}
  \end{titlepage}

%---------------------------------------------------------------------------------------------------
%---------------------------------------------------------------------------------------------------
%---------------------------------------------------------------------------------------------------
%--------------------------------------------------------------------------------------------------- 
\chapter*{Abstract}

  Parasitism can be defined as an interaction between species in which one of the interaction 
    partners, the parasite, lives in or on the other, the host. The parasite draws food from its 
    host and harms it in the process. According to estimates, above 40\% of all eukaryotes are 
    parasites. Nevertheless, it is computationally difficult to obtain information whether a 
    particular taxon is a parasite making it difficult to query large sets of taxa.

  Here we test in how far it is possible to use the Open Tree of Life (OTL), a synthesis of 
    phylogenetic trees on a backbone taxonomy (resulting in unresolved nodes), to expand available 
    information via phylogenetic trait prediction. We use the Global Biotic Interactions (GloBI) 
    database to categorise 25,992 and 34,879 species as parasites and free-living, respectively and 
    predict states for over $\sim$2.3 million (97.34\%) leaf nodes without state information.

  We estimate the accuracy of our maximum parsimony based predictions using cross-validation and 
    simulation at 60-80\% overall, while strongly varying between clades. The cross-validation 
    results in an accuracy of 98.17\% which is explained by the fact that the data are not uniformly 
    distributed. We describe this variation across taxa as associated with available state and 
    topology information. We compare our results with several smaller scale studies, which used 
    manual expert curation and conclude that computationally inferred state changes largely agree in 
    number and placement with those. In clades in which available state information is biased 
    (mostly towards parasites, e.g. in Nematodes) phylogenetic prediction is bound to provide 
    results contradicting conventional wisdom.

  This represents, to our knowledge, the first comprehensive computational reconstruction of the 
    emergence of parasitism in eukaryotes. We argue that such an approach is necessary to allow 
    further incorporation of parasitism as an important trait in species interaction databases and 
    in individual studies on eukaryotes e.g. in the microbiome.

%---------------------------------------------------------------------------------------------------

  % This study focuses on the ancestral state reconstruction of parasitism in the tree of life of
  %   Eukaryota. We predict unknown states of species and estimate origins and losses of parasitism. \\
  % The challenge here is the size of the tree and the little information about it. \\
  % Such a large phylogenetic tree does not completely exist and therefore we work with a synthesis 
  %   tree of OTL \cite{Hinchliff2015} which is highly multifurcated. \\
  % For the 2,535,437 leaf nodes we could not gather much data. From the GloBI database 
  %   \cite{Poelen2014} which we used, we could only collect 25,992 parasitic and 34,879 free-living 
  %   species. It follows that we have $\approx 2.4\%$ state information. \\
  % So far, especially small scale or highly manual studies have been carried out. In this scale, it 
  %   requires different data sources to be interconnected. \\
  % We performed an analysis of existing algorithms and selected a Sankoff maximum parsimony algorithm 
  %   using the R package \textit{Castor} \cite{Louca2017}. \\
  % Nevertheless, the results are convincing and even though this is a purely computational approach 
  %   which did not include human experts input, results coincide with prior knowledge. Also regarding 
  %   the number of events, our estimates coincide with previous results by human experts, e.g. the 
  %   study by Weinstein and Kuris \cite{Weinstein2016}. \\

\tableofcontents
\clearpage
\pagenumbering{arabic}

%---------------------------------------------------------------------------------------------------
%---------------------------------------------------------------------------------------------------
%---------------------------------------------------------------------------------------------------
%--------------------------------------------------------------------------------------------------- 
\chapter{Introduction}
  This thesis is about the analysis of ancestral state reconstruction algorithms for non-binary 
    trees, applied to the currently largest phylogenetic synthesis tree of Open Tree of Life (OTL)
    \cite{Hinchliff2015}, with the application of prediciton of parasitism. \\
  
  For about 50 years, people have been working on ancestral state reconstruction. The first paper 
    is by Camin and Sokal, who in 1965 were working on algorithms for discrete-state data 
    \cite{Camin1965}. Different methods have been developed and the question is which method is the 
    most suitable for the problem at hand: The ancestral state reconstruction for a huge 
    multifurcated tree with \colorbox{red}{binary/two discrete} states. \\
  Royer-Carenzi et al. distinguishes two major classes of ancestral state reconstruction methods: \\
  The first is maximum parsimony: explain the current state with the least number of state changes 
    between the child and its ancestor. \\
  The other class she presents describes modeling the character evolution as a stochastic process and 
    using the likelihoods to compute the possible ancestral character states. This is generally done 
    with a continuous time Markov model \cite{RoyerCarenzi2013}.

  Parsimony methods are simple methods in which you can not include information such as branch 
    length (development time of the species). This is also their main point of criticism compared to 
    the more complex likelihood based models. Since there are no branch lengths or other information 
    available for the present relationship tree, and we thus can not take advantage of the 
    other models, we decided in favor for parsimonious.

  We will test the existing maximum parsimony algorithms of Fitch \cite{Fitch1971} and Sankoff 
    \cite{Sankoff1975} for this task and estimate their predictive power. The present tree structure 
    of OTL is not binary. A tree is multifurcated if each node has multiple ($n >= 3$) children 
    \cite{Felsenstein2003}. \\
  Maximum parsimony methods are developed for phylogenies, which are usually depicted as binary trees.
    Parsimonious in phylogeny refers to favoring the tree that needs the least evolutionary change 
    to explain the observed data. In our case, it is about the change of states 'is free-living' or 
    'is parasitic'. \\
    % Parsimonious in this context means to minimize the number of transitions between species states 
    % (free-living or parasitic). 
  The Sankoff method is implemented by Louca et al. for the non-binary case and is available as an R 
    package called \textit{Castor} \cite{Louca2017}. In addition, we have implemented the Fitch 
    method and adapted it for multifurcated trees. \\% to make it usable and comparable for our current problem.
  This achieves that we can predict ancestral states and unknown states of living species for large 
    non-binary relatives trees.

  % For this purpose, various studied methods and their advantages and disadvantages will be compared. \\
  % These algorithms were developed especially for the binary problem, since one has considered much 
  %   smaller subtrees, of which one knows all the splits. \\
  % By developing a whole of an entire tree of life, the problem arises that it is by no means binary. \\

  To accomplish this task, a large phylogenetic tree and information about the current species states 
    is needed. \\
  The biggest 'phylogenetic tree' is a synthesis of phylogenetic trees filled with a taxonomic trees 
    given by Open Tree of Life (OTL) \cite{Hinchliff2015}. %This tree is not binary and therefore the 
    % developed algorithms are not directly applicable. Therefore, an algorithm must be found that can 
    % work with non-binary trees. \\
  For the information about the current states of the species we use the interaction database Global 
    Biotic Interactions (GloBI) \cite{Poelen2014}. The data in GloBI are stored as interactions e.g. 
    species A parasitize species B. From this we conclude that species A is parasitic and species B 
    free-living.

  At this point a few words to the term parasitic. There are different definitions. Since we use 
    GloBI to classify species, we use their definition of parasitism. Again, in GloBi, Ontobee 
    definitions are used \cite{Xiang2011}.
  The interaction \textit{has parasite} is defined as: "An interaction relationship between two 
    organisms living together in more or less intimate association in a relationship in which 
    association is disadvantageous or destructive to one of the organisms."\footnote{
      \hyperlink{http://www.ontobee.org/ontology/RO?iri=http://purl.obolibrary.org/obo/RO_0002445}
      {ontobee.org/ontology/RO?iri=http://purl.obolibrary.org/obo/RO\_000244}; Last checked: 22.03.2018.
    }. This definition includes: ecto- and endoparasites, parasitoids, kleptoparasites and pathogenes. \\

  The objectives of this work are the following points: (1) Find a suitable ancestral state 
    reconstruction method. (2) Perform reconstructing on the Eucaryotic synthesis tree of OTL. \\
  The goal of Point 1 is to evaluate the possible methods based on a simulation of our data 
    situation. \\
  The one by Louca et al. implemented Sankoff algorithm is the best in our comparisons. Therefore, 
    point 2 consists of reconstructing the ancestral states and predicting the unknown leaf states. 
    And then perform an evaluation of the results.

%---------------------------------------------------------------------------------------------------
%---------------------------------------------------------------------------------------------------
%---------------------------------------------------------------------------------------------------
%--------------------------------------------------------------------------------------------------- 
\chapter{Aims}
  The aim of this thesis is the application of maximum parsimony algorithms to non-binary 
    trees and very large data sets. In particular, the example find the 'Origins of Parasitism' 
    throughout the Eukaryota Tree of Life. \\
  For these large phylogenetic synthesis trees, however, ancestral state reconstruction has so far 
    only been done for Bacteria and Archaea for binary traits by Goberna and Verdú \cite{Goberna2015}.
    However, this differs from eukaryotes in the sense that complex traits such as parasitism depend
    on more than one gene. \\
  \todo{Warum ist das relevant? Was könnte man dann tun?}

%---------------------------------------------------------------------------------------------------
%---------------------------------------------------------------------------------------------------
%---------------------------------------------------------------------------------------------------
%--------------------------------------------------------------------------------------------------- 
\chapter{Methods}
  As stated in the introduction, in this thesis, a maximum parsimony algorithm is applied to the 
    whole tree of life to obtain an ancestral state reconstruction of free-living versus parasite 
    states. \\
  So far, these reconstructions have been made mainly on binary trees with better data availability. 
    Therefore, a simulation is first performed to evaluate existing algorithms and decide how they 
    may be adapted to our given problem. This is to perform the ancestral state reconstruction for a 
    multifurcated (non-binary) tree using binary states. \\
  Accordingly, in addition to the necessary data sets (GloBI, OTL), the chosen algorithm and the 
    evaluation of its results, this chapter also deals with the previously performed simulation and 
    the evaluation of the various algorithms and their parameters. \\
  Figure \ref{fig:workflow} briefly outlines these relationships. A more detailed view of the 
    workflow can be found in the appendix \ref{fig:BigWorkflow}.
  \begin{figure}[h!]
    \centering
    \includegraphics[width=0.6\textwidth]{Figures/Workflow-overview.png}
    \caption{The Workflow of the resulting procedure with the following steps: \\
      (1) Retrieve phylogenentic tree data as input for the tree (OTL) and the state information (GloBI).
      (2) Get metadata of these for a realisitc simulation of the maximum parsimony algorithms (Fitch \& Sankoff).
      (3) Build and run the simulation.
      (4) Evaluation of parameters for the simulation and the ancestral state reconstruction of the real tree.
      (5) Evaluate the accuracy of developed algorithms and choose the best.
      (6) Run the resulting algorithm on the original data.
      (7) Evalute and interprete the results. %$\rightarrow$ Origins etc...
    }
    \label{fig:workflow}
  \end{figure}
  
  %---------------------------------------------------------------------------------------------------
  %---------------------------------------------------------------------------------------------------
  %---------------------------------------------------------------------------------------------------
  \section{Description of data sets}
    Two types of data are needed for an ancestral state reconstruction: a tree and information about 
      the states. \\
    For the tree Open Tree of Life (OTL) is used \cite{Hinchliff2015} and for the state information 
      the Global Biotic Interactions database (GloBI) is taken \cite{Poelen2014}. \\
    % These are presented below.

    %---------------------------------------------------------------------------------------------------
    %---------------------------------------------------------------------------------------------------
    \subsection{OTL}
      For this project a large database for phylogenetic trees and also for a taxonomic tree is needed. 
        Since an ancestral state reconstruction algorithm is applied to the phylogenetic tree, and for 
        the assessment and other properties the taxonomy provides much more information. \\
      OTL gives a synthesis of phylogenetic trees (currently 819 trees) and a taxonomic tree\footnote{
          \hyperlink{https://tree.opentreeoflife.org/about/synthesis-release/v9.1}
          {https://tree.opentreeoflife.org/about/synthesis-release/v9.1}; Last checked: 22.03.2018.
        }. OTL also 
        includes the large phylogenetic database TreeBASE \cite{Hinchliff2015}. \\
      \todo{Das steht auf der Website nicht in dem Paper...} \\
      For phylogenetic data, there are at least five big data collections, namely:
      \begin{itemize}
        \item ITIS (Integrated Taxonomic Information System) \cite{ITIS}
        \item NCBI (National Center for Biotechnology Information) \cite{NCBI1988}
        \item WORMS (World Register of Marine Species) \cite{WoRMS2018}
        \item GBIF (Global Biodiversity Information Facility) \cite{GBIF}
        \item OTT (OpenTreeOfLife-Taxonomy) \cite{Hinchliff2015}
      \end{itemize}
      ITIS is only a small set of 100\% confirmed and named species. GBIF is not composed with the help 
        of phylogeny, the same is valid for the NCBI taxonomy. The WORMS taxonomy is a way too small 
        dataset of mostly marine species. \\
      Here the taxonomy from OTL is used because it is including most of the known taxonomies and is 
        synthesised by preffering taxonomies that match with available phylogenetic data. Furthermore 
        the team from OTL preferre a maximum number of species \cite{Hinchliff2015}. This is resulting 
        in somekind of hybrid between taxonomy and phylogeny. \anmerkungstext{Wie genau ist das ein 
        Hybrid? Genauer beschreiben, was Du damit meinst... (Thilo)} \\

      A closer look is being made to some of the features of the synthesis tree. On the one hand the 
        distribution of the taxa and on the other the distribution of the nodes on the taxa. Since this 
        is not directly relevant for this study, there is a section in the appendix 
        \ref{sec:otl analysis}.

    %---------------------------------------------------------------------------------------------------
    %---------------------------------------------------------------------------------------------------
    \subsection{GloBI}
      The most big interaction databases are offline or outdated. For example: IWDB (Interaction Web 
        Database) \cite{IWDB2003}, Webs on the Web \cite{WOW2004}, Animal Diversity Web \cite{Myers2003} 
        and ecoweb \cite{Cohen2010}. GloBI is including most of the known ones and is still growing 
        actively \cite{Poelen2014}. So the question was answered rather quickly which interaction 
        database could be used. \\

      This database consists of entries of the form: species A (source) interacts with B (target). \\
      A number of interactions have been identified\footnote{\hyperlink{
        https://github.com/jhpoelen/eol-globi-data/blob/master/eol-globi-lib/src/main/java/org/eol/globi/domain/InteractType.java
        }{https://github.com/jhpoelen/eol-globi-data/.../InteractType.java}; Last checked: 22.03.2018.}
        , including those that the species source or target has become a parasite or a free-living 
        species from the biological perspective. These are the following:
      \begin{itemize}
        \item free-living source: preysOn, eats, flowersVisitedBy, hasPathogen, pollinatedBy, 
          hasParasite, hostOf
        \item free-living target: preyedUponBy, parasiteOf, visitsFlowersOf, pathogenOf, hasHost
        \item parasite source: parasiteOf, pathogenOf
        \item parasite target: hasParasite, hasPathogen
      \end{itemize}
      Of these interactions, e.g. species A parasitize species B, the state of the species is 
        determined, here is species A parasitic and species B free-living. The case a parasite 
        conquers (parasitizes) another parasite yields conflicting states for the second species. 
        This is solved by preferring parasitic. \\
      For each species known IDs are stored in GloBI. This includes OTT (the taxonomy of OTL). All 
        species that have stored an OTT id and have a matching interaction are formed into two lists: 
        parasites and free-livings.

  %---------------------------------------------------------------------------------------------------
  %---------------------------------------------------------------------------------------------------
  %---------------------------------------------------------------------------------------------------
  \section{Metadata analysis}
    In order to generate a realistic simulation, influencing parameters are investigated. \\
    Since the transitions are minimized in an ancestral state reconstruction, this is an important 
      parameter to consider. On the other hand, the completeness of our input data are influencing 
      values. Therefore, two major types are distinguished:
    \begin{enumerate}
      \item Biological parameters (A result of the evolutionary process.):
        \begin{itemize}
          \item transition probabilities
        \end{itemize}
      \item Distribution of the loss of information:
        \begin{itemize}
          \item Loss of topology ($\rightarrow$ mutlifurcations)
          \item Unknown information about states of some leaf nodes
        \end{itemize}
    \end{enumerate}
    The influence of these parameters are tested on our result using our simulation (section 
      \ref{sec:methods - simulation}).
   
    %---------------------------------------------------------------------------------------------------
    %---------------------------------------------------------------------------------------------------
    \subsection{Transition probabilities}
      As mentioned above, in an ancestral state reconstruction, transitions are minimized, with 
        transition probabilities playing a role. \\
      This subsection deals with these transition probabilities from free-living (hereinafter / as a 
        formula FL) to parasitic (hereinafter P) and vice versa: $\mathcal{P}(FL \rightarrow P)$, 
        $\mathcal{P}(P \rightarrow FL)$. \\
      Different parasite types have different transition probabilities. It is very difficult to make a 
        statement about these probabilities. It is generally assumed in this work that there are 40\% 
        parasites and 60\% free-livings which is based on the estimates by Windsor \cite{Windsor1998} 
        and $\mathcal{P}(FL \rightarrow P) = \mathcal{P}(P \rightarrow FL)$, as a result of the 
        diversity of parasites and the lack of general determinations for this. These parameters are 
        debated in subsection \ref{sec:discussion - transition probabilities} of the discussion. \\

      For the maximum parsimony analysis of the real data, all transition probabilities are equated.
        However, the used castor package \cite{Louca2017} offers the possibility to enter different 
        transition probabilities.
      \begin{wrapfigure}{r}{0.5\textwidth}
        \begin{center}
          \includegraphics[trim = 0mm 0mm 0mm 0mm, clip, width=0.45\textwidth]{Figures/40-60.png}
        \end{center}
        \caption{Simulated distributions of the states: \\
          60\% Free-living - 40\% Parasites \\
          red: parasites, blue: free-living, \\
          the threshold is at 0.4}
        \label{fig:Beta distribution}
      \end{wrapfigure}

      In the simulation two beta distributions have been chosen and a threshold that indicates the change 
        between states. \\
      Different thresholds with different beta distributions are simulated, with different distributions 
        of parasites and free-livings:
        \begin{itemize}
          \item 50\% P to 50\% FL,
          \item 40\% P to 60\% FL,
          \item 30\% P to 70\% FL and 
          \item 20\% P to 80\% FL
        \end{itemize}
        (\todo{see results simulation, ref...}). Figure \ref{fig:Beta distribution} 
        shows one example of these. \\

    %---------------------------------------------------------------------------------------------------
    %---------------------------------------------------------------------------------------------------
    \subsection{Missing information} \label{sec:methods - missing information}

      A binary tree with $n$ leaf nodes has $n-1$ internal nodes. The present Eukaryota tree of OTL has 
        2,293,463 leaf nodes and only 41,974 internal nodes, that is:
      $$100-\frac{100}{(2293463-1) \times 41974} \approx 98.16 \%$$
        missing internal nodes. This means that there is a lack of information about the underlying 
        phylogeny. Instead of a binary tree this tree is highly multifurcated. \\

      % \anmerkungstext{Nested models were compared using likelihood ratio tests, models using different 
      %   predictors were compared according to their deviance and AIC. (Emanuel)} \\

      For the present Eukaryota tree with 2,293,463 leaf nodes, 34,869 free-livings and 25,962 parasites 
        are found, which are
        $$100-\frac{100}{2293463 \times (34860+25962)} \approx 97.34 \%$$
        unknown states of leaf nodes. \\

      In the simulation, the influence of the multifurcation and missing data in leaf nodes on the 
        predictive accuracy of the ancestral state reconstruction algorithms is tested. \\
      For the real data, generalized linear models are compared with poisson respectively binomial 
        regression according to their residuals. In order to compare models of different complexity, the 
        BIC (Bayesian Information Criterion) values were calculated in addition to the residuals.\\
      There are two different information criteria: AIC (Akaike Information Criteria) and BIC. \\
      The advantage of the BIC is that the penalty is dependent on the sample size and is therefore 
        advantageous for large samples. \\

      For each node, depth, min, max and mean height were noted. Where the depth of a node is the 
        distance (number of edges) to the root node and the height of a knot is described as the 
        largest distance to a leaf node. In this work, a distinction is made between minimum, maximum 
        and average distance ($\rightarrow$ min, max and mean height). \\
      The influence in the modeling of these parameters was tested, additive as well as multiplicative. \\

      For all these calculations, the following R functions were used: \textit{glm()}, \textit{anova()} 
        and \textit{BIC()}.
      % https://stat.ethz.ch/R-manual/R-devel/library/stats/html/glm.html

  %---------------------------------------------------------------------------------------------------
  %---------------------------------------------------------------------------------------------------
  %---------------------------------------------------------------------------------------------------
  \section{Ancestral state reconstruction methods} \label{sec:ancestral state reconstruction}
    As presented in the introduction, there are some methods for ancestral state reconstruction. For 
      this purpose, various studied methods and their advantages and disadvantages are compared below. \\
  
  % The first methods were only brute force \cite{Camin1965} 
  %   Next came a set of parsimony algorithms such as: Fitch-parsimony \cite{Fitch1971}, 
  %   Wagner-parsimony \cite{Swofford1987} ... \todo{weitere?}. \\
  %   With more and more data, there is now the possibility to use more information to calculate the 
  %   probabilities of the ancestral states. In addition to the states of the leafs, algorithms could 
  %   also use branch lengths. The likelihood based algorithms came more in interest. \\
  
    Royer-Carenzi et al. distinguishes two major classes of ancestral state reconstruction methods: \\
    The first is to explain the current state with the least number of state changes between an 
      ancestor and his child, this is called parsimonious. \\
    The other class she presents involves modeling the character evolution as a stochastic process and 
      using the likelihoods to compute the possible ancestral character states. This is generally done 
      with a continuous time Markov model \cite{RoyerCarenzi2013}. \\
    \todo{Pasqualin et al. unterscheiden noch eine weitere Methode: stochastic mapping...} \\

    One of the major disadvantages of parsimony methods is that, unlike likelihood approaches, they 
      can not take divergece times (branch length) into account. Since the OTL does not include 
      development times of species, this is irrelevant. \\
    Another problem pointed out by Royer-Carenzi is that parsimony approaches are either based on 
      predefined parameters (generalized parsimony) or on strong and often controversial assumptions, 
      like irreversibility of transitions for dollo parsimony. Again, this problem is unimportant to 
      the problem at hand, because in the analysis of the entire Eukaryota tree only generalized 
      models make sense. \\

    Parsimony-based methods are used in this work, since they are fully sufficient for the presented 
      use case here. Following the principle of the simpler model first. \\
    \anmerkungstext{Nochmal darstellen warum und selbst wenn sie fully sufficient wäre, heißt es ja 
      nicht, dass man sie nehmen muss. Also: motiviere! (Renard)} \\
    Felsenstein \cite{Felsenstein2003} discusses in his book two algorithms that generalize all 
      previous methods (from Camin and Sokal \cite{Camin1965}, \todo{Kluge and Farris} and Farris 
      \cite{Farris1970}): Fitch parsimony \cite{Fitch1971} and Sankoff parsimony \cite{Sankoff1975}. \\
    \anmerkungstext{Unter Farris war auch noch der Begriff Wagner trees in gebrauch, als 
      Verallgemeinerung der parsimonious trees von Camin und Sokal. (Lydia)} \\
    \todo{Wagner-parsimony \cite{Swofford1987}} \\
    
    Thus, the methods used in this work are those of Fitch and Sankoff. For Fitch, the algorithm has 
      been extended from binary to multifurcated trees. For the Sankoff algorithm, Louca and Doebeli 
      have presented an implementation for non-binary trees published in an R package named 
      \textit{castor} \cite{Louca2017}.

    %---------------------------------------------------------------------------------------------------
    %---------------------------------------------------------------------------------------------------
    \subsection{Fitch maximum parsimony}
      Fitch maximum parsimony is an algorithm for rooted, binary trees and describes an ancestral state 
        reconstruction for discrete states \cite{Fitch1971} by minimizing transitions between states. \\
      Note, the original Fitch algorithm has the sole purpose of minimizing the number of transitions 
        and not reconstructing the ancestral nodes. Felsenstein \cite{Felsenstein2003} describes a 
        simple extension for the reconstruction. Cunningham et al. \cite{Cunningham1998} have refined 
        these. \todotext{Wir haben mit ein paar kleinen änderungen optimiert... und schließlich auf 
        multifurcated angepasst...} \todotext{eigentlich ist Cunningham 'nur' eine kritische 
        Neubewertung. Sie beziehen ihren Algortihmus auf Swofford und Maddison...} \\
      To understand the differences to the multifurcated case, the algorithm for the binary case is 
        briefly explained and referred to the extension. \\

      Input: A rooted, binary tree, with state informations in the leaf nodes. Each node is depicted as 
        a set of states. There are only two states in this thesis, free-living (FL) and parasitic (P). 
        Internal nodes have three sets, which are empty at the beginning, excluding the root node, it 
        has only one. Leaf nodes have their state as a set (e.g. \{FL\} or \{P\}, unknown leaf nodes the 
        union of all possible states (\{FL, P\}). \\

      The algorithm traverses three times through the tree and fills these sets. \\
      In each step, two sets are considered and their intersect formed. There are two cases:
      \begin{enumerate}
        \item The intersection is not empty and corresponds to the new set.
        \item The intersection is empty. $\rightarrow$ Build the union of these sets as new set.
      \end{enumerate}
      \begin{wrapfigure}{r}{0.5\textwidth}
        \centering
        \includegraphics[width=0.4\textwidth]{Figures/Fitch1.png}
        \caption{Fitch algorithm for binary trees. \\
          The unknown leaf node is discribed with both states. Computed internal nodes (exclusive the 
          root node) consists of three sets, where the last set is the final one (bold). \\
          From the second internal node (seen from the root node) there are several possibilities to 
          create the second and third set.}
        \label{fig: binary Fitch}
      \end{wrapfigure}
      First traverse from the leaf nodes to the root / move down the tree / postorder tree traversal.
      
        Each internal node is formed from its child nodes, where at the beginning the only information
        lies. \\
      Second traverse from the root node to the leafs. Each internal node is formed from its father node 
        and its sibling node. \\
      Last traversion (direction does not matter): Build the final state for every node. It is formed 
        from the sets of previous traversals. \\
      (The original Fitch algorithm was designed to minimize transitions without predicting actual states 
        of internal nodes, so it was just the first traversal.) \\
      The extension to the non-binary case is quite obvious, but holds some opportunities. In this case, 
        more than two children may be present for the first traversal, but the incision or union may 
        also be formed over more than two sets. Also in the second traversing, there may be several 
        sibling nodes. However, there are several possibilities here that were all tested and compared 
        in the simulation. Some of these options are already available in the binary case:
      \begin{itemize}
        \item The father node has (except for the root node) two state sets, because he came through 
          the up-traversing previously. Are both sets used or only the first traversing?
        \item Since there are several siblings, do \colorbox{red}{you} first of all make the cut or union, 
          or directly in the whole with the father node?
      \end{itemize}
      The first point already has an effect on the binary case. Figure \ref{fig: binary Fitch} shows 
        both possibilities of the three sets. \\
      Cunningham uses only the first state set of the father node \cite{Cunningham1998}. \\
      From these two points four different versions of Fitch were formed:
      \begin{enumerate}
        \item Fitch 1: First state set of father node; intersection/union of siblings first.
        \item Fitch 2: First state set of father node; intersection/union of siblings together with father node.
        \item Fitch 3: Both state sets of father node; intersection/union of siblings first.
        \item Fitch 4: Both state sets of father node; intersection/union of siblings together with father node sets.
      \end{enumerate}
      \begin{figure}
        \centering
        \includegraphics[width=0.8\textwidth]{Figures/simulation_fitch_evaluation.png}
        \caption{Test of Fitch Versions.}
        \label{fig:Fitch versions}
      \end{figure}
      These four versions were tested in the simulation with 100 trees and 10000 leaf nodes and a
        distribution of 60\% FL to 40\% P. Figure \ref{fig:Fitch versions} shows this over all unknown 
        node percentage. \\
      % At 90\% unknown nodes and 90\% of multifurcation of the internal nodes, version 1 was 89.26\%, 
      %   version 2 was 89.26\%, version 3 was 89.35\%, and version 4 was 89.31\% correct. Therefore, 
      %   only version 3 was used for all further simulations.
      At 95\% unknown nodes and 95\% of multifurcation of the internal nodes, version 1 was 88.37\%, 
        version 2 was 88.37\%, version 3 was 88.4\%, and version 4 was 88.39\% correct. Therefore, 
        only version 3 was used for all further simulations.
        
    % Fitch versions:
    % | 88.37 % | 88.37 % | 88.4 % |88.39 % |


    %---------------------------------------------------------------------------------------------------
    \subsubsection{Sankoff}
      Maximum parsimony algorithm from Sankoff implemented in the R package castor \cite{Louca2017}. \\
      \todo{transition probabilities: all equal}

  %---------------------------------------------------------------------------------------------------
  %---------------------------------------------------------------------------------------------------
  %---------------------------------------------------------------------------------------------------
  \section{Simulation}\label{sec:methods - simulation}
    The simulation compares these different of ancestral state reconstruction algorithms with each 
      other. \\
    First different implementations of the Fitch maximum parsimony are compared and then the best of 
      them is compared with the implementation of the Sankoff algorithm of the \textit{Castor} package 
      \cite{Louca2017}. \\

    \begin{figure}[h!]
      \centering
      \includegraphics[width=1\textwidth]{Figures/Workflow-Simulation.png}
      \caption{A simulation was performed to compare different maximum parsimony algorithms. \\
        The course of the simulation with influence of the metadata analysis from the real data can
          be seen: \\
        (1) A phylogenetic tree is created randomly. (2) Simulate node states for all nodes. (3) 
          'Forget' internal states and some leaf node states. (4) 'Lose' phylogeny information. (5) 
          Make a nodelist for the algorithm. (6) Run algorithms. (7) Evaluate results. \\
        Points 3 and 4 are influenced by metadata of the real-data analysis.}
      \label{fig:Simulation Workflow}
    \end{figure}
    The course of a simulation is shown in Figure \ref{fig:Simulation Workflow}. The individual steps 
      are explained in the following. \\

    A tree is needed to do a simulation of ancestral state reconstruction. It had to be decided whether 
      to take the real tree or simulate a tree. In this simulation, trees are created randomly, as 
      one can replicate a complete binary phylogentic tree. Thus, there is also the possibility to 
      simulate the multifurcation. \\
    To get a random binary tree, the Phylo package from biopython is used \cite{Cock2009}. They offer 
      a \textit{randomized()} function which returns a BaseTree\footnote{
        \hyperlink{https://github.com/biopython/biopython/blob/master/Bio/Phylo/BaseTree.py}
        {https://github.com/biopython/biopython/blob/master/Bio/Phylo/BaseTree.py}
      }. The credibility of this choice is examined in subsection 
        \ref{sec:discussion - simulate phylogeny} of the discussion.\\

    The next step is to simulate states for all nodes. \\
    The root node is defined as ancestor of all subsequent species and in this case, determined to be
      free-living. Therefore, a beta distribution for free-living is used at the beginning. Now 
      traverse from the root to the leaf nodes, always pulling out of the current distribution until 
      the randomly drawn number is above the threshold and the new node changes state. Figure 
      \ref{fig:set node states} shows a part of this simulating states. \\
    To ensure that the parameter of the binomial distribution is restricted to the [0,1] interval, it 
      is modeled with a beta distribution as in Figure \ref{fig:set node states}. \\

    After traversing through the tree, each state is saved in a nodelist associated with the  
    \begin{wrapfigure}[21]{r}{0.65\textwidth}
      \centering
      \includegraphics[width=0.65\textwidth]{Figures/40-60_all.jpg}
      \caption{Set node states: Distribution of states (left); traversion through the tree (right). \\
        Start with a free-living root node (FL: red). \\
        (1) + (2) Draw random numbers for its children from the free-living distribution (red), the numbers are under the threshold 
        $\rightarrow$ the nodes are again free-living; go on with the children. \\
        (3) The number drawn is above the threshold. $\rightarrow$ The node state changes to parasitic (P: blue). \\
        (4) Now draw random numbers from the parasite distribution (red) until one number lies unter the threshold. Then change back.}
      \label{fig:set node states}
    \end{wrapfigure}
    node ID which is the OTT from OTL. \\

% simulate loss of information
    Here begins the simulation of the lost information. This is on the one hand the states and on the 
      other the topology of the tree. \\

    In the real tree, there is usually only information about species living today $\rightarrow$ leaf 
      nodes. And beyond only a small percentage of these. All information about the states of the 
      internal node and one leaf node is 'forgotten' and stored in another column to the node. \\

    Last step for the preparation is the multifurcation of the tree. As previously explained, some 
      divisions in the tree are not known, so the real tree is not binary. This multifurcation is 
      simulated by an equally distributed percentage of forgotten internal nodes. \\
    
    Different percentages of forgetting information are simulated, see figure 
      \ref{fig:influence of unknown data} in the results. \\
  
    The last step is the evaluation of the results. This is done with a simple difference calculation 
      of the node states. \\
    In the nodelist, the originally simulated states and the newly calculated states are stored for 
      each node ($FL = 0$, $P = 1$). The sum of the differences of the node states gives the distance 
      of the prediction to the original tree.

  %---------------------------------------------------------------------------------------------------
  %---------------------------------------------------------------------------------------------------
  %---------------------------------------------------------------------------------------------------
  \section{Real data analysis}
    \todo{Evaluation section in methods?}
    \begin{itemize}
      \item Import tree
      \item Import interactions
      \item run castor algorithm / and others?
      \item interprete results (leave-100-out cross-validation)
    \end{itemize}
    We then evaluated the results in two ways. On the one hand, we have examined the results of some 
      subtrees for their credibility with the help of biological background knowledge. On the other 
      hand, we performed a leave-100-out cross-validation.

  %---------------------------------------------------------------------------------------------------
  %---------------------------------------------------------------------------------------------------
  %---------------------------------------------------------------------------------------------------
  \section{Implementation}
    The complete code is located on GitHub: 
      \hyperlink{github.com/Irallia/IZW-HU-Parasites}{github.com/Irallia/IZW-HU-Parasites}. \\
    Most of the code was written in Python. The analyzes and the use of the Castor package in R. There 
      are some shell scripts to execute whole workflows.

%---------------------------------------------------------------------------------------------------
%---------------------------------------------------------------------------------------------------
%---------------------------------------------------------------------------------------------------
%---------------------------------------------------------------------------------------------------
\chapter{Results}
  This work deals with the ancestral state reconstruction of the entire Eukaryota relatives tree. \\

  For this reconstruction, we first analyzed our data. This is the tree of OTL \cite{Hinchliff2015}
    and the data for the leaf nodes (free-living or parasite states) from GloBI \cite{Poelen2014} 
    (section \ref{sec:results - missing information}). \\
  Next we compared different possible methods. We decided to take a closer look at maximum parsimony 
    algorithms because they are best suited to the problem at hand. The reason for this can be found 
    in section \ref{sec:ancestral state reconstruction}. We tested these on different simulations of 
    the data and compared their predictive power (section \ref{sec:results - simulation}). \\
  We used the Sankoff algorithm \cite{Sankoff1975} implemented by Louca et al. \cite{Louca2017}, 
    which performs best, for the actual reconstruction on the real data. \\

  We then evaluated the results in two ways. On the one hand, we have examined the results of some 
    subtrees for their credibility with the help of biological background knowledge (subsection 
    \ref{sec:results - biological view}). On the other hand, we performed a leave-100-out cross-validation
    (Subsection \ref{sec:results - cross-validation}) and predicted approximately 98.17\% states 
    correct.

  
%   Otherwise, this chapter is mainly about the actual reconstruction of the states. This means, on 
%     one hand investigation of origins and losses of the inner nodes and on the other, the prediction 
%     of unknown states of leaf nodes. \\
%    
  %---------------------------------------------------------------------------------------------------
  %---------------------------------------------------------------------------------------------------
  %---------------------------------------------------------------------------------------------------
  \section{Metadata analysis - Missing information} \label{sec:results - missing information}
    As previously presented in the methods (section \ref{sec:methods - missing information}), we have 
      two types of missing information: unknown states of leaf nodes and multifurcation. \\

    A tree is multifurcated if there are nodes that have more than two children. In the case of a 
      phylogenetic tree, the ridge of multifurcation describes the amount of lack of information about 
      the topology of the tree. A complete phylogenetic tree would be binary, meaning that the number 
      of leaf nodes equals the number of internal nodes minus one. Since we only work with a 
      artificially constructed tree (synthesis of several phylogenetic trees), this tree is strongly 
      multifurcated: It has 241,974 internal nodes for 2,293,463 leaf nodes $\rightarrow$ 98.16 \% 
      missing nodes. \\
    We calculated these percentages of missing nodes and also missing state information for some 
      subtrees and plotted them in table \ref{table:percentage loss information subtrees}. \\
    \begin{table}[h!]
      \begin{center}
        \begin{tabular}{ |l||r|r| }
          \hline
          \bfseries Subtree of & \bfseries Unknown States & \bfseries Multifurcation \\ 
          \hline \hline
          Eukaryota       & 97.34\%  & 98.16\% \\
          \hline \hline
          Metazoa         & 96.44\%  & 87.93\% \\ \hline
          Fungi           & 98.87\%  & 96.97\% \\ \hline
          Chloroplastida  & 99.14\%  & 89.46\% \\
          \hline \hline            
          % Arthropoda      & 97.49\%  & 89.95\% \\ \hline
          Apicomplexa     & 86.26\%  & 87.16\% \\ \hline
          Nematoda        & 89.01\%  & 88.59\% \\ \hline
          Chordata        & 88.59\%  & \cellcolor{green!50}66.49\% \\ \hline
          Platyhelminthes & \cellcolor{green!50}68.73\%  & 80.34\% \\
          \hline \hline            
          Insecta         & 97.11\%  & 90.78\% \\
          \hline  
        \end{tabular}
      \end{center}
      \caption{Examination of subtrees regarding missing information. \\
        The percentage values show the ratio of missing information of: unknown states (missing state 
          information of leaf nodes) and multifurcation (missing internal nodes). \\
        The subtrees are from different taxa: domain (Eukaryota), kingdom (Metazoa, Fungi, 
          Chloroplastida), phylum (Apicomplexa, Nematoda, Chordata, Platyhelminthes) and class 
          (Insecta). \\
        The two by far smallest values were highlighted in green.} 
      \label{table:percentage loss information subtrees} 
    \end{table}
    
    We also have for a first overview we collected for each node its number of children (degree $-1$), 
      and plotted this in two histograms, see figure \ref{fig:childrenOfNodes}. \\
    \begin{figure}
      \centering
      \begin{subfigure}[b]{0.59\textwidth}
        \includegraphics[width=0.9\textwidth]{Figures/multifurc.pdf}
        \caption{Histogram with automatic binwidth. \\ ~}
      \end{subfigure}
      \begin{subfigure}[b]{0.4\textwidth}
        \includegraphics[trim = 0mm 0mm 30mm 0mm, clip, width=0.9\textwidth]{Figures/multifurc_small.pdf}
        \caption{Histogram with $binwidth = 1$. Higher multifurcation than 30 has been cut off. \\ light blue: binary; \\ dark blue: multifurcation}
      \end{subfigure}
      \caption{Histograms about the multifurcation of the internal nodes of the synthesis tree. \\
        For each node, the number of children (degree $-1$) was collected. A node is multifurcated 
        if it has more than two children, so we deducted two from each number of children. We have 
        plotted this in two histograms: number of children $-2$ on the x-axis with log scale and the 
        number of nodes with this amount on the y-axis.}
      \label{fig:childrenOfNodes}
    \end{figure}
    The multifurcation affects only the internal nodes. We collected the number of children $-2$ of 
      every node (a node with two children is binary). That means it describes the number of nodes 
      which we have lost from the real (binary) phylogenetic tree. \\
    It can be recognized that we are very far from a binary tree. \\

    %---------------------------------------------------------------------------------------------------
    %---------------------------------------------------------------------------------------------------
    \subsection{Data artifacts}
      At this point we also found out that there are nodes with only one child node (55,700 nodes). \\
      The most of these nodes are next to a leaf node, others are deep in the tree (3,956 with height 
        $>2$). They are probably a result from the fact that taxonomic information has been incorporated 
        into a phylogeny. \\
      \anmerkungstext{probably? Da ist eine Spekulation hier? Kann man das nicht herausfinden? 
        Spekulieren solltest du am ehesten erst in der Diskussion. (Thilo)} \\
      Some examples of these nodes are:
      \begin{itemize}
        \item Nephroselmidophyceae: (class) \\
          https://tree.opentreeoflife.org/opentree/argus/ottol@1038762
        \item Phrynocrinidae: (family) \\
          https://tree.opentreeoflife.org/opentree/argus/ottol@3647979
        \item Elaeocarpus sylvestris: \\
          https://tree.opentreeoflife.org/opentree/argus/opentree9.1@ott166969
      \end{itemize}

    %---------------------------------------------------------------------------------------------------
    %---------------------------------------------------------------------------------------------------
    \subsection{Poisson regression of the multifurcation}
      For a regression analysis of the multifurcation first set up a simple generalized linear model and 
        then output the intercept of it:
      \lstset{language=R}
      \begin{lstlisting}[gobble=6]
        kingdom.furc.mod <- glm(multifurc~kingdom, 
                                    data=inner.taxa,
                                        family="poisson")
        # get the intercept:
        summary(kingdom.furc.mod)
      \end{lstlisting}
      Since the multifurcation can also be 0, the intercept describes the expected mean value and thus, 
        if it is greater than 0, it says that we have a multifurcation: $2.62 > 0$ $\Rightarrow$ there 
        is a multifurcation. \\
      We next compared the intercept of the different kingdoms and found that
        $$1.22 \text{ (Fungi intercept)} > -0.11 \text{ (Metazoa intercept)} > -0.09 \text{ (Chloroplastida intercept)}$$
        % Coefficients:
        %                                         Estimate Std. Error z value Pr(>|z|)
        % (Intercept)                             2.624897   0.003641  721.00   <2e-16
        % kingdom['ott352914', 'Fungi']           1.224650   0.004054  302.10   <2e-16
        % kingdom['ott361838', 'Chloroplastida'] -0.090993   0.003960  -22.98   <2e-16
        % kingdom['ott691846', 'Metazoa']        -0.114873   0.003732  -30.78   <2e-16
        that means the strength of the multifurcation \colorbox{red}{is the same/has the same size 
        sorting}. \todo{kann das sein?} \\

      We investigated several models that could describe multifurcation: \\
      In doing so, we allowed the different influence of the taxa and the heights and depths of a node 
        to be included. From this we got 9 times 4 models of different complexity levels (see first row 
        of table \ref{table:BIC multifurcation}). \\
      We first compared the models based on their residuals. These give us the error of the model. If 
        the value is small, our data will be well modeled. These values are meaningful only for the same 
        complexity, therefore we then calculated the BIC (Bayesian information criterion) values of the 
        models and used them to evaluate the models. \\
      % \lstset{language=R}
      % \begin{lstlisting}[gobble=6]
      %   anova(kingdom.furc.mod, phylum.furc.mod, class.furc.mod, 
      %                 order.furc.mod, family.furc.mod, test="LRT")
      %   BIC(kingdom.furc.mod)
      %   ...
      % \end{lstlisting}
      The deviance table with the residuals can be found in the appendix (table 
        \ref{table:Residuals multifurcation}). Table \ref{table:BIC multifurcation}. lists the BICs.

      \begin{table}[h]
        \begin{center}
          \begin{tabular}{ |l|r|r|r|r| }
            \hline
            \bfseries Model / Taxa & \bfseries Kingdom & \bfseries Phylum & \bfseries Class & \bfseries Order \\
            \hline \hline
            multifurc $\sim$ taxa & 8273333 & \cellcolor{green!15}7937828 & \cellcolor{green!20}7842157 & \cellcolor{green!30}7644249 \\
            multifurc $\sim$ taxa & 8257680 & \cellcolor{green!15}7922207 & \cellcolor{green!20}7826490 & \cellcolor{green!35}7574154 \\
            \hline
            multifurc $\sim$ taxa + depth & 8273318 & \cellcolor{green!15}7934322 & \cellcolor{green!20}7839364 & \cellcolor{green!35}7539999 \\
            multifurc $\sim$ taxa + max.height & \cellcolor{green!15}7993515 & \cellcolor{green!25}7749121 & \cellcolor{green!30}7661817 & \cellcolor{green!40}7416211 \\
            multifurc $\sim$ taxa + min.height & 8251211 & \cellcolor{green!20}7875521  & \cellcolor{green!25}7778327 & \cellcolor{green!35}7516883 \\
            multifurc $\sim$ taxa + mean.height & \cellcolor{green!20}7825417 & \cellcolor{green!30}7644249 & \cellcolor{green!35}7572474 & \cellcolor{green!45}7340741 \\
            \hline
            multifurc $\sim$ taxa * depth & 8235932 & \cellcolor{green!20}7836755 & \cellcolor{green!25}7757688 & \cellcolor{green!45}7383808 \\
            multifurc $\sim$ taxa * max.height & \cellcolor{green!15}7963438 & \cellcolor{green!30}7693555 & \cellcolor{green!30}7614820 & \cellcolor{green!45}7335338 \\
            multifurc $\sim$ taxa * min.height & 8214030 & \cellcolor{green!20}7808940 & \cellcolor{green!30}7690618 & \cellcolor{green!45}7336627\\
            multifurc $\sim$ taxa * mean.height & \cellcolor{green!25}7768360 & \cellcolor{green!35}7536296 & \cellcolor{green!50}7484953 & \cellcolor{green!50}7206369 \\
            \hline
          \end{tabular} 
        \end{center}
        \caption{BIC (Bayesian information criterion) values of the multifurcation models. \\
          These models were created with the R function \textit{glm()} and compared with the 
            \textit{BIC()} function. This results in the listed BIC values.}
        \label{table:BIC multifurcation} 
      \end{table}
      % * Residuals: Fehler - wieviele Werte sind nicht gut modelliert. (umso kleiner umso besser - grün) \\
      Within the every complexity class it can be seen that the mean height gives the best additional factor. \\
      Despite higher complexity, the BIC values are getting smaller from model to model, meaning that 
        the finest model available here is also the best one of these. Lower taxa than orders (e.g. 
        family) were computationally too expensive to calculate. \\
      The model \textit{multifurc $\sim$ order * mean.height} turns out to be the best of our models, 
        whereby it is possible that e.g. \textit{multifurc $\sim$ family * mean.height} is better.

    %---------------------------------------------------------------------------------------------------
    %---------------------------------------------------------------------------------------------------
    \subsection{Binomial regression of the unknown state information}

      Next to the problem of the multifurcation of the tree is the less of interaction data we have for 
        the species. For the ancestral state reconstruction, we need information about the states 
        (free-living or parasite) in the leaf nodes. \\
      The eukaryotic synthesis tree has 293,463 leaf nodes. The GloBI database has 5,346,414 interactions 
        (at 29.01.2018). Out of this data we got 51,337 distinct free-living species and 47,332 
        distinct parasite species for the whole tree of life. From the Eukaryota we could determine 
        25,992 and 34,879 species as parasites and free-living. With 2,293,463 leaf nodes we still have 
        about 97.34\% unknown leaf nodes. In the discussion, section 
        \ref{sec:discussion - data situation}, we will talk about ways to extend this data. \\
      
      We also compared different models in terms of their BICs (Table: \ref{table:BIC unknown information}). 
        The Residuals are not very meaningful here, since all models have different complexities. % /dimensions
        For the sake of completeness, the associated deviance tables are located with the residuals in 
        the appendix \ref{sec:Residuals unknown information}.

      \begin{table}[h!]
        \begin{center}
          \begin{tabular}{ |l|r|r|r|r| }
            \hline
            \bfseries Model / Taxa & \bfseries Kingdom & \bfseries Phylum & \bfseries Class & \bfseries Order \\
            \hline \hline
            multifurc $\sim$ taxa & \cellcolor{green!15}545799 & \cellcolor{green!35}500004 & \cellcolor{green!45}485121 & \cellcolor{green!45}484681 \\
            \hline
            multifurc $\sim$ taxa + depth & \cellcolor{green!15}544862 & \cellcolor{green!40}493808 & \cellcolor{green!45}481869 & \cellcolor{green!50}478851 \\
            \hline
            multifurc $\sim$ taxa * depth & \cellcolor{green!15}544179 & \cellcolor{green!45}489845 & \cellcolor{green!45}481494 & \cellcolor{green!50}478188 \\
            \hline
          \end{tabular} 
        \end{center}
        \caption{BIC (Bayesian information criterion) values unknown state information models. \\
          These models were created with the R function \textit{glm()} and compared with the 
            \textit{BIC()} function. This results in the listed BIC values.}
        \label{table:BIC unknown information} 
      \end{table}
      It also follows from this table that the most complex model is the best. In general, the BIC 
        values are smaller than those of the multifurcation models. The modeling here is thus better. \\
      Again, the calculation of finer models (e.g. order or family) was too expensive. \\

      These missing data modeling results can be used to better simulate the data.

  %---------------------------------------------------------------------------------------------------
  %---------------------------------------------------------------------------------------------------
  %---------------------------------------------------------------------------------------------------
  \section{Results of simulation / Influence of different parameters} \label{sec:results - simulation}
  
    As presented, we compare two methods in our simulation to their prediction accuracy: Fitch and 
      Sankoff. \\
    Figure \ref{fig:influence of unknown data} shows the results of examining various parameters.
    \begin{figure}[h!]
      \centering
      \includegraphics[trim = 0mm 100mm 0mm 0mm, clip, width=\textwidth]{Figures/simulation_evaluation_1.png}
    \end{figure}
    ~ \\ ~ \\ ~ \\
    \begin{figure}
      \centering
      \includegraphics[trim = 0mm 0mm 0mm 150mm, clip, width=\textwidth]{Figures/simulation_evaluation_1.png}
      \caption{Influence of unknown data to prediction. \\
        The first column describes the distributions of free-livings and parasites with a given 
          threshold for the respective simulations to the right. \\
        The middle column investigates the influence of the unknown states, the right the influence of
          the strength of the multifurcation. \\
        The y-axes indicate the percentage of correctly predicted states (including known states). On 
          the x-axis the percentage of forgotten states or missing internal nodes. \\
        Each point corresponds to the average of one hundred simulations, each with 10,000 leaf nodes. \\
        For the middle column we set the strength of the multifurcations to 0.95\% similar to the 
          real data and in the right column the amount of the unknowns to 0.95\% also similar to the 
          real data.}
      \label{fig:influence of unknown data}
    \end{figure}
    First, we tested different transition probability distributions including threshold (first column). 
      It can be observed that the more balanced the percentages of free-livings and parasites, the 
      worse the prediction of the algorithms (second and third column). \\
    It can be seen that both algortithms are always over 50\% prediction and therefore better than 
      guessing. Moreover, they are usually close to each other, with Sankoff always makes better 
      predictions except for equally distributed states as Fitch. \\
    On the other hand, we examined the influence of missing internal nodes (ridge of multifurcation) 
      and missing leaf node information (unknown leaf nodes). \\
    Both factors have a relatively linear influence on the Sankoff method. Fitch, on the other hand, 
      breaks significantly in his prediction from about 70\% unknown leaf nodes or mutifurcation. \\

    Since we have a lot of missing data in most subtrees (both internal nodes and state informations), 
      the Sankoff gives a better prediction and was thus used for real data analysis.

  %---------------------------------------------------------------------------------------------------
  %---------------------------------------------------------------------------------------------------
  %---------------------------------------------------------------------------------------------------
  \section{Results of the real data analysis created with Sankoff}
    This section is about evaluating the prediction of real data using the Sankoff method. \\
    It is divided into two subsections. \ref{sec:results - biological view}: The analysis of some subtrees using 
      biological background knowledge. \ref{sec:results - cross-validation}: Presentation of the results 
      of the leave-100-out cross-validation.

    %---------------------------------------------------------------------------------------------------
    %---------------------------------------------------------------------------------------------------
    \subsection{Biological view} \label{sec:results - biological view}
      To analyze the results, we have selected some phyla (subtrees) to evaluate our results 
        selectively from the biological point of view: Chordata, Nematoda, Platyhelminthes and 
        Apicomplexa. \\
      In Table \ref{table:phylum leaf nodes states} we compare the given states with the predicted ones. \\
      Several factors such as the distribution of existing input data via parasitics and free-livings, 
        faulty input data from GloBI and reinforcement of errors by multifurcation play an important 
        role, which crystallize out in these examples. \\
      An important factor here is that the credibility of the results. The accuracy of the input data 
        stands and falls with the presence and the correctness of the data of GloBI. Errors of incorrect 
        input data can be amplified by incorrect prediction of unknown species and can be reversed in 
        order to improve the data situation of GloBI. \\
      Since we look at such large trees we can not expect to know all the parasites, so we look at 
        individual positives. This is positive in the sense that the majority have the opposite state.

      \begin{table}
        \begin{center}
          \begin{tabular}{ |l|r||r|r||r|r|r|r| }
            \hline
            & & \multicolumn{2}{c||}{\bfseries original states} & \multicolumn{4}{c|}{\bfseries final states} \\
            \bfseries Phylum & \bfseries \# nodes & \bfseries FL & \bfseries P & \bfseries 0 \bfseries (FL) & \bfseries 0.3 &\bfseries  0.5 & \bfseries 1 (P) \\
            \hline \hline
            Chordata & 91785        & 10451 & 18          & 91759 & 0 & 0 & 26 \\
            &                       & 99.83\% & 0.49\%  & 99.97\% & & & 0.03\% \\ \hline
            Nematoda & 30127        & 21 & 3289           & 1604 & 142 & 1196 & 27185 \\
            &                       & 0.63\% & 99.37\%  & 5.32\% & 0.47\% & 3.97\% & 90.23\% \\ \hline
            Platyhelminthes & 22683 & 7 & 7086            & 175 & & & 22508 \\
            &                       & 0.1\% & 99.9\%    & 0.77\% & 0 & 0 & 99.23\% \\ \hline
            Apicomplexa & 1863      & 1 & 255             & 1 & 0 & 0 & 1862 \\
            &                       & 0.39\% & 99.61\%  & 0.05\% & & & 99.95\% \\
            % \hline \hline
            % Phylum & \# nodes & FL & P
            %   & 0 (FL) & 0.4 & 0.5 & 0.67 & 0.75 & 1 (P) \\
            % Arthropoda & 1198981 & 18912 & 11141 
            %   & 1099509 & 1313 & 22478 & 4176 & 1665 & 70223 \\
            % & & 62.93\% & 37.07\%
            %   & 91.7\% & 0.11\% & 1.87\% & 0.35\% & 0.14\% & 5.86\% \\
            \hline
          \end{tabular} 
        \end{center}
        \caption{Some selected phyla (subtrees), only leaf nodes, were examined for the evolution 
          of the amount and percentage of given data to predicted data.}
        \label{table:phylum leaf nodes states} 
      \end{table}

      In contrast to the other phyla examined, the phylogeny in the Chordata is more pronounced (less 
        multifurcation) (see Table \ref{table:percentage loss information subtrees}). This results in 
        less variance of errors. What is reflected in the results from the table 
        \ref{table:phylum leaf nodes states}. There are 18 parasites as input data and only 8 more are 
        predicted. The Chordata mostly consist of free-living species, so this seems believable. We 
        started with 99.83\%  species and predict 99.97\% species as free-living (including already 
        known nodes).

      \begin{figure}
        \centering
        \includegraphics[trim = 0mm 0mm 0mm 0mm, clip, width=0.85\textwidth]{Figures/ChordataParasites.png}
        \caption{Parasites of Phylum: Chordata. \\
          All parasite data of the chordata were mapped into a rough taxonomy (phylum, class, order, 
            family) in order to understand their distribution and affiliation. \\
          The internal nodes are the wanted taxa from OTL (with the addition of \# input parasites to 
            $\rightarrow$ \# predicted parasites). \\
          The leaf nodes are the input parasites (green) and the predicted parasites (white $\rightarrow$ 
            green).}
        \label{fig:ChordataParasites}
      \end{figure}

      We mapped the few parasitic species in a \colorbox{red}{rough/simple} taxonomy (see Figure \ref{fig:ChordataParasites}): \\
      Known parasitic birds belong to the order Sauria. Here we know from Rothschild  here there are 
        breeding parasites, like the cuckoo and clepto-parasites as the skuas \cite{Rothschild1957}. We
        got 6 input parasites from GloBI and there are no predictions: A woodpecker - 
        \textit{Sphyrapicus varius} and a duck - \textit{Aix sponsa}, a cow bird - \textit{Molothrus ate} 
        known as broodparasite and some others. \\
      An example of the amplification of mistakes here are the carp. There is a paper from which GloBI 
        concludes: Grass carp (\textit{Ctenopharyngodon idella)} has Pathogen common carp 
        (\textit{Cyprinus carpio})\footnote{
          \hyperlink{https://www.globalbioticinteractions.org/?interactionType=hasParasite&targetTaxon=Cyprinus\%20carpio}
          {https://www.globalbioticinteractions.org/?interactionType=hasParasite\&targetTaxon=Cyprinus\%20carpio}
        }. Since there is hardly any information about free-living, it follows that all siblings are 
        also predicted to be parasitic. \\
        % Other birds: Columba livia, Nestor meridionalis, Sphyrapicus varius \\
        % Snake: Storeria dekayi (probably caused by wrong input data.)
      % \begin{itemize}
      %   \item Order Acanthomorphata(3):, one slime fish (family Blenniidae) and two others without family
      %   \item Order Anura (1): a toad (Family Bufonidae) - there exists some parasitic species in here, 
      %     they eat fin or pieces of larger fish.
      %   \item Order Siluriformes (1): a catfish (Family Loricariidae) - there exists some parasitic 
      %     species in here.
      % \end{itemize}

      The Apicomplexa are a parasitic phylum. We found only one input organism: Stemonitis fusca as a 
        free-living species. In GloBI it is listed as being parasitized by Nectria candicans and Nectriopsis 
        sporangiicola\footnote{
          \hyperlink{https://www.globalbioticinteractions.org/?interactionType=parasiteOf&targetTaxon=Stemonitis\%20fusca}
          {https://www.globalbioticinteractions.org/?interactionType=parasiteOf\&targetTaxon=Stemonitis\%20fusca}
        }. The algorithm has not predicted new free-livings. \\

      Most species of Platyhelminthes (flatworms) are parasites, although there are also free-living, 
        predatory feeding species. These are summarized in the Turbellaria, while the parasites are 
        divided into three other classes \cite{Ax1961}. This also corresponds to our observations. There 
        is one class (Rhabditophora) that contains all but one single exception of free-living species 
        of this phylum, which includes the Turbellaria. \\
      It should be noted, however, that this classification is outdated, as it has been proven that 
        the Turbellaria are not monophyletic. But we will not go into that here. \\
      For the Platyhelminthes we had more state information for the leaf nodes compared to the other 
        considered subtrees, see Table \ref{table:percentage loss information subtrees}. We start with 
        0.1\% free-livings and predicted 0.77\% as free-living species. \\

      With the Nematoda it looks more complicated. Large parts of Nematoda are free-living, but we 
        found only 5.32\% of them. Blaxter et al. estimates the order of 25,000 parasites in the 
        Nematoda \cite{Blaxter2015} and speaks of at least seven independently arosed parasitism 
        \cite{Blaxter1998}. In a recent article Blaxter identifies 18 origins \cite{Blaxter2015} in 
        Nematoda. \\
      The problem at this point, however, is: Hallan speaks of the fact that only 23,000 species 
        were described by the Nematoda but is it assumed that one million or more species are in this 
        phylum.\footnote{J. Hallan, unpublished; 
          \hyperlink{https://insects.tamu.edu/research/collection/hallan/}
          {https://insects.tamu.edu/research/collection/hallan/}
        }. \todo{Link ist nicht erreichbar!} The parasites have been much more studied and thus we start 
        with only 0.63\% (table \ref{table:phylum leaf nodes states}) free-living species. Against such 
        a shifted data situation, the algorithm is almost powerless to make correct predictions. And yet 
        the percentage has increased to 5.34\%. \\

      % \anmerkungstext{Das sind schonmal vier große Kontraste, wenn dann noch Zeit bleibt, die 
      %   schwirigen... Arthropoden, Fungi, Pflanzen... (Emanuel)} \\

      \todotext{Im Folgenden folgen 3 weitere ähnliche tabellen. Einmal eine ähnliche Tabelle inklusive 
        interner Knoten \ref{table:Phylum internal nodes}, zweimal die Übersicht über die Kingdoms 
        Blattknoten \ref{table:Kingdom leaf nodes} bzw interne Knoten \ref{table:Kingdom internal nodes}. Welche 
        nehmen wir? Rest Appendix oder ganz raus?}

      \begin{table}[h!]
        \begin{center}
          \begin{tabular}{ |l|r||r|r||r|r|r|r|r|r| }
            \hline
            & & \multicolumn{2}{c||}{\bfseries original states} & \multicolumn{6}{c|}{\bfseries final states} \\
            \bfseries Phylum & \bfseries \# nodes & \bfseries FL & \bfseries P
              & \bfseries 0 (FL) & \bfseries 0.4 & \bfseries 0.5 & \bfseries 0.67 & \bfseries 0.75 & \bfseries 1 (P) \\
            \hline \hline
            Chordata & 122546 & 10451 & 18 
              & 122473 & 0 & 0 & 0 & 0 & 73 \\
            Nematoda & 33564 & 21 & 3289 
              & 846 & 0 & 1133 & 0 & 0 & 31585 \\
            Platyhelminthes & 27142 & 7 & 7086 
              & 1010 & 0 & 175 & 0 & 0 & 25957 \\
            Apicomplexa & 2102 & 1 & 255 
              & 1 & 0 & 0 & 0 & 0 & 2101 \\
            % \hline
            % Arthropoda & 1319460 & 18912 & 11141 
            %   & 1207204 & 1313 & 25499 & 4852 & 1957 & 78635 \\
            \hline
          \end{tabular}
        \end{center}
        \caption{Some selected phyla (subtrees), including internal nodes, were examined for the evolution 
        of the amount and percentage of given data to predicted data.}
        \label{table:Phylum internal nodes}
      \end{table}

      \begin{table}[h!]
        \begin{center}
          \hspace*{-2cm}\begin{tabular}{ |l|r||r|r||r|r|r|r|r|r|r|r| }
            \hline
            & & \multicolumn{2}{c||}{\bfseries original states} & \multicolumn{8}{c|}{\bfseries final states} \\
            \bfseries Kingdom & \bfseries \# nodes & \bfseries FL & \bfseries P
              & \bfseries 0 (FL) & \bfseries 0.25 & \bfseries 0.33 & \bfseries 0.4 & \bfseries 0.5 & \bfseries 0.67 & \bfseries 0.75 & \bfseries 1 (P) \\
            \hline \hline
            none & 75446 & 45 & 529 
              & 13426 & 220 & 24082 & 0 & 7792 & 5302 & 0 & 24493 \\
            Fungi & 31457 & 577 & 2983
              & 38520 & 0 & 0 & 0 & 5723 & 0 & 0 & 266463 \\
            Chloroplastida & 416478 & 3519 & 77
              & 410795 & 0 & 0 & 0 & 4182 & 0 & 0 & 1501 \\
            Metazoa & 1491012 & 30758 & 22373
              & 1328135 & 0 & 0 & 930 & 25535 & 4423 & 1665 & 130324 \\
            \hline  
          \end{tabular}
        \end{center}
        \caption{All kingdoms (subtrees), only leaf nodes, were examined for the evolution of the 
          amount and percentage of given data to predicted data.}
        \label{table:Kingdom leaf nodes}
      \end{table}

      \begin{table}[h!]
        \begin{center}
          \hspace*{-2cm}\begin{tabular}{ |l|r||r|r||r|r|r|r|r|r|r|r| }
            \hline
            & & \multicolumn{2}{c||}{\bfseries original states} & \multicolumn{8}{c|}{\bfseries final states} \\
            \bfseries Kingdom & \bfseries \# nodes & \bfseries FL & \bfseries P
              & \bfseries 0 (FL) & \bfseries 0.25 & \bfseries 0.33 & \bfseries 0.4 & \bfseries 0.5 & \bfseries 0.67 & \bfseries 0.75 & \bfseries 1 (P) \\
            \hline \hline
            none & 84456 & 45 & 529 
              & 15035 & 243 & 25910 & 0 & 8764 & 6183 & 0 & 28140 \\
            Fungi & 324105 & 577 & 2983
              & 39088 & 0 & 0 & 0 & 5858 & 0 & 0 & 274803 \\
            Chloroplastida & 460457 & 3519 & 77
              & 454211 & 0 & 0 & 0 & 4688 & 0 & 0 & 1558 \\
            Metazoa & 1670956 & 30758 & 22373
              & 1485749 & 0 & 0 & 1313 & 29002 & 5102 & 1957 & 147833 \\
            \hline  
          \end{tabular}
        \end{center}
        \caption{All kingdoms (subtrees), inkluding internal nodes, were examined for the evolution of the 
        amount and percentage of given data to predicted data.}
        \label{table:Kingdom internal nodes}
      \end{table}

      %---------------------------------------------------------------------------------------------------
      \subsubsection{Origins and Losses} \label{sec:results - origins and losses}

        Weinstein and Kuris have been searching for origins of parasitism in Animalia \cite{Weinstein2016}. 
          They identified 223 parasitic origins: 223 in Metazoa $\supset$ 143 in Arthropoda $\supset$ 87 
          in Insecta. \\
        This has led us to count the origins and losses of parasitism in our investigation as well. \\
        We count only one origin / loss in a parent node with different children's nodes. \\
        Here we have encountered a problem: The Castor algorithm gives us probabilities for states. That 
          means there are also nodes with state like 0.3 or 0.5. The question is, how to count these state 
          changes. We decided to round the values.
        % \begin{lstlisting}[gobble=6]
        %   possible states: 0, 0.333, 0.4, 0.5, 0.667, 0.75, 1
        %   rounded to:    0  0      0    0    1      1     1
        % \end{lstlisting}

        \begin{table} [h]
          \begin{center}
            \begin{tabular}{ |l|r|r||r|r|r| }
              \hline
              \bfseries Domain  & \bfseries \# internal & \bfseries \# leaf & \bfseries Rootnode & \multicolumn{2}{c|}{\bfseries without and with rounding} \\ 
              \bfseries / Kingdom  & \multicolumn{2}{c||}{\bfseries nodes} & \bfseries state & \bfseries \# origins & \bfseries \# losses \\
              \bfseries / Phylum / Class & & & & \bfseries (FL -> P) & \bfseries (P -> FL) \\
              \hline \hline
              Eukaryota & 241974 & 2293463 & 1.0 & 415 & 363 \\
              & & & P & 462 & 369 \\
              \hline \hline
              Metazoa & 179944 & 1491012 & 0.5 & 294 & 123 \\
              & & & & 321 & 129 \\ \hline
              Fungi & 9534 & 314571 & 0.5 & 80 & 222 \\
              & & & & 97 & 222 \\ \hline
              Chloroplastida & 43486 & 412434 & 0.0 & 40 & 2 \\
              & & & FL & 42 & 2 \\
              \hline \hline            
              Arthropoda & 120479 & 1198981 & 0.0 & 260 & 102 \\
              & & & FL & 281 & 108 \\ \hline
              Apicomplexa & 239 & 1863 & 1.0 & 0 & 1 \\
              & & & P & 0 & 1 \\ \hline
              Nematoda & 3437 & 30127 & 1.0 & 0 & 11 \\
              & & & P & 2 & 11 \\ \hline
              Chordata & 30761 & 91785 & 0.0 & 12 & 1 \\
              & & & FL & 12 & 1 \\ \hline
              Platyhelminthes & 4459 & 22683 & 1.0 & 0 & 5 \\
              & & & P & 0 & 5 \\
              \hline \hline            
              Insecta & 91256 & 989572 & 0.0 & 234 & 77 \\
              & & & FL & 245 & 77 \\ 
              \hline  
            \end{tabular}
          \end{center}
          \caption{Origins and losses}
          \label{table:origins and losses} 
        \end{table}

        In Table \ref{table:origins and losses} we can see, that we found some more origins than Weinstein 
          and on top of that some losses. \\

        When more closely regarding the same phyla as in the section before: Chordata, Nematoda, 
          Platyhelminthes and Apicomplexa. \\
        Chordata are full of free-living species and so we see only a few origins of parasitism. The root
        and mostly all species are predicted as free-living. \\
        In Apicomplexa and the Platyhelminthes also look credible. Our algorithm gives us only one loss of
          parasitism in Apicomplexa and five in the Platyhelminthes. They are both from the root over
          mostly all species predicted as parasites. \\
        Nematoda is again full of problems. The root node is predicted as a parasite and so we have more 
          losses of parasitism for the less information of free-living species in this phylum. The rest 
          is parasitic \\
        As we have already mentioned Blaxter et al. found at least seven origins of parasitism 
          \cite{Blaxter1998}. If we assume that the root node of Nematoda is free-living, then some 
          losses would have to turn around and become Origins. So it could be that we end up in a similar 
          size as Blaxter.
      
    %---------------------------------------------------------------------------------------------------
    %---------------------------------------------------------------------------------------------------
    \subsection{Leave-100-out cross-validation} \label{sec:results - cross-validation}
      For a further validation of our results, we carried out a leave-100-out cross-validation. \\
      In order to achieve about 15\% of the 60,871 input node states with a validation, 
        we randomly left out 100 states 100 times. To make smaller amounts up to leave-one-out had too 
        much computational effort. \todo{warum 15\%?}\\
        
      Of these 10,000 nodes, 9,238 were unique. From the unique nodes, we predicted approximately 
        98.17\% correct and thus 1.82\% wrong, with duplicate draws always having the same prediction 
        (data not shown). \\
      
      We have again dealt with the question of how this data is best modeled. The influence of the taxa 
        (kingdom, phylum, class) and the depth of the leaf nodes was re-modeled and the BICs compared 
        (table \ref{table:BIC cross-validation}). Lower taxa than classes (e.g. order) were 
        computationally too expensive to calculate. \\
      The Residuals are included in the appendix (table \ref{table:Residuals cross-validation}) for the 
        sake of completeness; here too, due to the different complexity, they do not have much meaning 
        for the comparison of the models. \\

      \todo{table for these numbers?}
      % 100/(25992+34879)*10000 = 16.43
      % 100/(25992+34879)*9238 = 15.18
      % 10000 nodes 9238 unique
      % 9060 nodes $\approx 98.17\%$ correctly predicted
      % 169 ($\approx 1.82\%$) wrongly predicted

      \begin{table}[h]
        \begin{center}
          \begin{tabular}{ |l|r|r|r| }
            \hline
            \bfseries Model / Taxa & \bfseries Kingdom & \bfseries Phylum & \bfseries Class \\% & Order \\
            \hline \hline
            correctly predicted $\sim$ taxa & 117936 & \cellcolor{green!40}112242 & \cellcolor{green!50}111733 \\% & XXXXX \\
            \hline
            correctly predicted $\sim$ taxa + depth & 117776 & \cellcolor{green!50}111304 & \cellcolor{green!50}111273 \\% & XXXXX \\
            \hline
            correctly predicted $\sim$ taxa * depth & 117709 & \cellcolor{green!50}111262 & \cellcolor{green!30}113135 \\% & XXXXX \\
            \hline
          \end{tabular} 
        \end{center}
        \caption{BIC (Bayesian information criterion) values of cross-validation prediction models. \\
          These models were created with the R function \textit{glm()} and compared with the 
            \textit{BIC()} function. This results in the listed BIC values.}
        \label{table:BIC cross-validation} 
      \end{table}

      The BIC values this time did not prove that the finest model is the best. Of our calculated models, 
       \textit{correctly predicted $\sim$ phylum * depth} has the smallest value. \\

      We examined the influence of the omitted data on the prediction. On average, about twice as many 
        leaf nodes are predicted differently. \color{red}The variance is very high. This means that 
        we have a high degree of dispersion and thus a stochastic situation exists. It also describes 
        the width of the present probability function. The standard deviation is about five times as 
        high as the number of omitted nodes, so the variability is quintupled. \color{black} Table 
        \ref{table:statistics cross-validation} shows these results. \\
      \begin{table}[h!]
        \begin{center}
          \begin{tabular}{ |cl||r|r|r|r|r| }
            \hline
            & & \bfseries min & \bfseries max & \bfseries mean & \bfseries variance ($\sigma^2$) & $\sigma$ \\
            \hline \hline
            \multirow{3}{*}{\bfseries distance} & \bfseries all     & 0 & 3587.70 & 224.96 & 313650.61 & 560.05 \\
            & \bfseries leaf nodes                                  & 0 & 3021.12 & 208.69 & 248103.38 & 498.10 \\
            & \bfseries internal nodes                              & 0 & 566.58 & 16.28 & 4927.95 & 70.20 \\ \hline
            \multirow{3}{*}{\bfseries lost} & \bfseries all states  & 100 & 100 & 100 & 0 & 0 \\
            & \bfseries FL states                                   & 44 & 66 & 57.25 & 19.50 & 4.42 \\
            & \bfseries P states                                    & 34 & 56 & 42.75 & 19.50 & 4.42 \\
            \hline
          \end{tabular}
        \end{center}
        \caption{Statistics about the leave-100-out cross-validation \\
          The distance between original and new states was calculated using the Euclidean metric. This 
            was summed over all states / all leaf node states and all internal node states. \\
          The lower half of the table describes the distribution of the 'lost' states between parasites 
            (P) and free-livings (FL).}
        \label{table:statistics cross-validation}
      \end{table}

%---------------------------------------------------------------------------------------------------
%---------------------------------------------------------------------------------------------------
%---------------------------------------------------------------------------------------------------
%---------------------------------------------------------------------------------------------------
\chapter{Discussion}
  This chapter deals with the question of how trustworthy our result of the ancestral state 
    reconstruction of the Eukaryota tree is. And further on how well our simulation can simulate the 
    real problem and thus make statements about the predictive power of the applied Sankoff algorithm.

  We have pursued this question in various investigations and yet, of course, further possibilities 
    for improvement remain. Despite these possibilities for improvement, our reconstruction gives a 
    first good assessment of the whole tree.

  \todo{Why is the study important?} \\
  \todo{How does this study relate to previous studies?}
  
  We evaluated our ancestral state reconstruction and the relational state prediction using three 
    methods: a simulation, a selective analysis from a biological perspective and a leave-100-out 
    cross-validation.
  
  Each theme has its own limitations. \\
  There are various expansion options and points of criticism for the simulation, which in 
    particular revolve around the question of how well we can simulate our real data. We dedicate 
    this question to the subsection \ref{sec:discussion - simulation}. \\
  Above all, the 'biological view' gives a first impression of the credibility of the results. But 
    he can not make any statistical statements. Here the interested researcher may have to analyze 
    the subtrees of relevant to him for himself. \\
   Of course, cross-validation can be expanded. With more computing power, a leave-1-out cross 
    validation could also be performed. However, our result on this point yields a strong result 
    with 98.17\% correct prediction that this does not seem necessary. \\
  One last limitation also exists in the data situation itself. We have pointed out in various 
    sections that we have a strong multifurcation and only a few data about leaf node states. In 
    figure \ref{fig:influence of unknown data} you can see how the predictive power can improve if 
    we have more data available. We discuss this point in section 
    \ref{sec:discussion - data situation}.

  % Wieviele Origins haben wir gefunden, was bedeutet diese Zahl? \\

  %---------------------------------------------------------------------------------------------------
  %---------------------------------------------------------------------------------------------------
  %---------------------------------------------------------------------------------------------------
  \section{Data situation} \label{sec:discussion - data situation}
    The used eukaryotic synthesis tree from OTL \cite{Hinchliff2015} has 241,974 internal nodes and 
      293,463 leaf nodes. In addition, we could specify 25,992 parasitic and 34,879 free-living 
      species from GloBI \cite{Poelen2014}. \\
    This gives us a high number of missing internal nodes (high multifurcation) and a low number of 
      node states of the leaf nodes. In table \ref{table:percentage loss information subtrees} we have 
      written down this percentage including some subtrees. \\

    The investigation of the taxonomy revealed that the OTL tree has three kingdoms: Chloroplastida, 
      Metazoa, Fungi, 53 phyla, 195 classes and 924 orders. \\
    Since the analysis of the tree is not part of this work, it should be mentioned here that, 
      according to recent findings, this is not complete and we lack some taxa in every rank. For 
      example, Cavalier-Smith says that one distinguishes between seven and nine kingdoms 
      \cite{CavalierSmith1981}. \\
    In \colorbox{red}{section/page} \pageref{subsec:listPhyla} of the appendix is a list of all phyla. \\

    Also, the global database is not infallible. We found out (section \colorbox{red}{x}) that there is some 
      misinformation. \\
    We were able to report some of these directly to Globi.

    We found also 57,352 (not distinct!) source species and 809,993 (not distinct!) target species 
      without OTT ids. Since we currently use only OTT ids, we could not use this information. \\
    At this point there is thus the possibility to use more of the existing data, if one performs a 
      matching with the other Ids.

  %---------------------------------------------------------------------------------------------------
  %---------------------------------------------------------------------------------------------------
  %---------------------------------------------------------------------------------------------------
  \section{Simulation} \label{sec:discussion - simulation}
    The aim of the simulation was to test the influence of various unknown or unsafe parameters in 
      order to test the credibility of the prediction. \\
    Some of these influences could not be tested: Firstly, the distribution of parasites and 
      free-livings in the tree produces various parameters that we could not test all (see subsection 
      \ref{sec:discussion - transition probabilities}) and, secondly, we discuss the simulation of a 
      phylogenetic tree on which the evaluation based on certain parts (see subsection 
      \ref{sec:discussion - simulate phylogeny}).

    %---------------------------------------------------------------------------------------------------
    %---------------------------------------------------------------------------------------------------
    \subsection{Transition probabilities} \label{sec:discussion - transition probabilities}
      We assume that different parasite types have different transition probabilities. Establishing a 
        common distribution across the eukaryotic tree is therefore difficult. \\
      Based on the estimates of Windsor \cite{Windsor1998}, we have assumed a distribution of 40\% 
        parasites to 60\% free livings in this work. As a result of the diversity of parasites and the 
        lack of generalizations, we have generally stated that 
        $\mathcal{P}(FL \rightarrow P) = \mathcal{P}(P \rightarrow FL)$. But it is also reasonable to 
        assume that in general $\mathcal{P}(FL \rightarrow P) > \mathcal{P}(P \rightarrow FL)$, because 
        a reverse mutation is usually less likely. However, one would have to determine how much this 
        difference is and thus discuss another parameter. \\
      
      In the simulation, we tested different beta distributions with different thresholds and found that 
        this combination has a considerable influence on the predictive power of the algorithms. \\
      At this point it would be possible to test other distributions with equal threshold values or 
        different threshold values with equal distributions. \\
      However, if you choose a strong free-living distribution and balance it with the threshold to 
        achieve a 40:60 parasite free-living distribution, for example, this could be very 
        computationally expensive. \\
      
      With these further simulations you could find out if the issue of distributions plays a big role. 
        Conversely, one could estimate possible distributions based on the data location in the tree. 
        However, this is likely to be very difficult given the poor data. For this reason, we have 
        decided in this work to accept very general values and not to speculate much. \\

    %---------------------------------------------------------------------------------------------------
    %---------------------------------------------------------------------------------------------------
    \subsection{Simulation of a phylogeny} \label{sec:discussion - simulate phylogeny}
      In our simulation, we start with the simulation of a phylogeny and then depict our data situation. \\
      For this first step we use the \textit{randomized()} function of the phylo package from biopython
        \cite{Cock2009}. The question is how close is such a randomized tree to a phylogeny. \\
      
      The problems that can arise are that different species develop at different rates. This means that 
        there could be sections in the tree with many branches (for example, subtrees of unicellular 
        species), ie with high depths and opposite sections (of very complex species, for example). It 
        can thus be assumed that the tree is not balanced. \\
      For more precise statements, one would have to take deeper undercurrents, which was not possible 
        in the context of this work.

  %---------------------------------------------------------------------------------------------------
  %---------------------------------------------------------------------------------------------------
  %---------------------------------------------------------------------------------------------------
  \section{Conclusion}

    From the simulation, we can conclude that we predict correctly about 60\% of the nodes in the 
      present data situation (see section \ref{sec:results - simulation}).
    The leave-100-out cross-validation even showed that we predict the omitted nodes to be 98.17\%  
      percent correct (see section \ref{sec:results - cross-validation}). \\
    This allows for the assumption that the data is grouped and not uniformly distributed and thus 
      smaller subtrees are present in which data are to be found in the simulation with smaller 
      multifurcation and smaller value for unknown nodes. This is also confirmed by our biological 
      analysis of the Chordata subtree (see subsection \ref{sec:results - biological view}). \\
    However, this means that the ancestral states' data in the direction of root node are probably 
      particularly unbelievable. Which makes the localization of Origins direction root node 
      difficult. The question remains, how much this affects the estimation of the number of origins. 
      Our comparison with the paper from Weinstein and Kuris \cite{Weinstein2016} (see subsection 
      \ref{sec:results - origins and losses}), however, leaves us with optimism.

%---------------------------------------------------------------------------------------------------
%---------------------------------------------------------------------------------------------------
%---------------------------------------------------------------------------------------------------
\bibliography{bibliographie}

%---------------------------------------------------------------------------------------------------
%---------------------------------------------------------------------------------------------------
%---------------------------------------------------------------------------------------------------
\chapter{Appendices}
  %---------------------------------------------------------------------------------------------------
  %---------------------------------------------------------------------------------------------------
  \section{Methods overview}
    % \begin{figure}[h!]
    %   \centering
    %   \includegraphics[angle=90, width=0.4\textwidth]{Figures/Workflow.png}
    %   \caption{Big overview of the whole Workflow}
    %   \label{fig:BigWorkflow}
    % \end{figure}
    \begin{figure}[h!]
      \centering
      \includegraphics[width=\textwidth]{Figures/Workflow.png}
      \caption{Big overview of the whole Workflow}
      \label{fig:BigWorkflow}
    \end{figure}

  %---------------------------------------------------------------------------------------------------
  %---------------------------------------------------------------------------------------------------
  \section{OTL analysis}\label{sec:otl analysis}

    %---------------------------------------------------------------------------------------------------
    \subsection{List of all phyla}\label{subsec:listPhyla}

    Phyla (53): \\
    Acanthocephala, Amoebozoa, Apicomplexa, Arthropoda, Ascomycota, Bacillariophyta, Basidiomycota, 
      Brachiopoda, Bryozoa, Chaetognatha, Chlorophyta, Chordata, Chromerida, Chytridiomycota, 
      Ciliophora, Cnidaria, Colponemidia, Ctenophora, Cycliophora, Echinodermata, Entoprocta, 
      Entorrhizomycota, Euglenida, Foraminifera, Gastrotricha, Glomeromycota, Gnathostomulida, 
      Haplosporida, Haptophyta, Hemichordata, Kinorhyncha, Loricifera, Microsporidia, Mollusca, 
      Myzostomida, Nematoda, Nematomorpha, Nemertea, Onychophora, Orthonectida, Phaeophyceae, 
      Picozoa, Placozoa, Platyhelminthes, Porifera, Priapulida, Rhodophyta, Rhombozoa, Rotifera, 
      Streptophyta, Tardigrada, Xanthophyceae \\
    Wobei von Streptophyta -> Anthocerotophyta, Marchantiophyta, Bryophyta, Tracheophyta als
      Phylum im Phylum gefunden und nicht einbezogen wurden und Magnoliophyta als Phylum in 
      Tracheophyta ebenfalls nicht. \\

    %---------------------------------------------------------------------------------------------------
    \subsubsection{Distribution of Taxa}
    - In the tree we can distinguish 28 different Taxa with the OTL taxonomic tree. \\
    - The most of them are hardly represented. The major taxonomic groups are: ... \\
    - Here \colorbox{red}{you} can see some characteristics of the Multifurcation of the tree. \\
    \begin{figure}[h!]
      \centering
      \includegraphics[width=0.9\textwidth]{Figures/TaxaTable2.JPG}
      \caption{Distribution of Nodes in Rank-Cathegories}
      \label{fig:taxaTable2}
    \end{figure}
    In a phylogeny, the taxonomic division of the tree is far too coarse, meaning that there should 
      be more subtaxa or 'unranked' nodes. But the closer we get to the root, the more the pure
      taxonomic tree is reflected. If the tree were binary, the taxa would have to double. But the 
      multipliers for some are much bigger and for others much smaller, which \colorbox{red}{you} can see in in figure 
      \ref{fig:taxaTable2}. \\
      ... (see Table \ref{table:...})
    
    \begin{table}[h!]
      \begin{center}
        % \begin{tabular}{ |>{\rowmac}l|>{\rowmac}r|<{\clearrow} }
        \begin{tabular}{ |l|r||r|r|r| }
          \hline
          \bfseries Taxa & \bfseries Number of Nodes & \bfseries Internal Nodes & \bfseries Leaf Nodes & \bfseries Terminal Leaf Nodes \\
          \hline \hline
          \setrow{\bfseries}domain & 1 & 1 & &  \\ \hline
          \setrow{\bfseries}kingdom & 3 & 3 & &  \\
          subkingdom & 2 & 2 & & \\
          infrakingdom & 5 & 5 & & \\
          superphylum & 2 & 2 & & \\ \hline
          \setrow{\bfseries}phylum & 57 & 54 & 3 & \\
          subphylum & 22 & 20 & 2 & \\
          infraphylum & 2 & 2 & & \\
          superclass & 8 & 8 & & \\ \hline
          \setrow{\bfseries}class & 209 & 175 & 34 & \\
          subclass & 120 & 106 & 14 & \\
          infraclass & 14 & 11 & 3 & \\
          superorder & 40 & 37 & 3 & \\ \hline
          \setrow{\bfseries}order & 1014 & 772 & 242 & \\
          suborder & 285 & 222 & 63 & \\
          infraorder & 95 & 78 & 17 & \\
          parvorder & 14 & 12 & 2 & \\
          superfamily & 829 & 626 & 203 & \\ \hline
          \setrow{\bfseries}family & 8449 & 6143 & 2306 & \\
          subfamily & 3090 & 2250 & 840 & \\
          supertribe & 1 & 0 & 1 & \\
          tribe & 1764 & 1285 & 479 & \\
          subtribe & 435 & 359 & 77 & \\ \hline
          \setrow{\bfseries}genus & 164656 & 105452 & 59204 & \\
          subgenus & 1266 & 869 & 397 & \\
          section & 3 & 2 & 1 & \\
          subsection & 1 & 1 & 0 & \\
          species group & 365 & 288 & 77 & \\
          species subgroup & 123 & 100 & 23 & \\ \hline
          \setrow{\bfseries}species & 2247251 & 7423 & 2239828 & 2228993 \\
          subspecies & 22437 & 75 & 22362 & 22239 \\
          variety & 1755 & 7 & 1748 & 1726 \\
          varietas & 5970 & 19 & 5951 & 5909 \\
          forma & 1181 & & 1181 & 1181 \\
          \hline \hline
          no rank & 954 & 719 & 235 & 7 \\
          no rank - terminal & 37452 & & 37452 & 37452 \\
          (no entry) & 40099 & 40099 & & \\
          \hline  
        \end{tabular}
        \caption{\todo{...}}
        \label{table:...} 
      \end{center}  
    \end{table}
    extended leaf nodes (real leaf nodes)

  %---------------------------------------------------------------------------------------------------
  \subsubsection{Distribution of data in the taxa}
    Mithilfe des taxonomischen Baums von OTL haben wir die Knoten ihren Kingdoms, Phyla und Classes 
      zugeteilt (see Table \ref{table:...}). \\

    \begin{table}[h]
      \begin{center}
        \begin{tabular}{ |l|r|l|r|r| }
          \hline
          \bfseries Kingdom (3) & \bfseries Number of Nodes & \bfseries Phylum (25) & \bfseries Number of Nodes & \bfseries max max height \\
          \hline \hline
          Metazoa & 1 465 207 & Arthropoda & 1 170 539 & 54 \\
          & & Chordata & 106 650 & 74 \\
          & & Mollusca & 80 022 & 22 \\
          & & Platyhelminthes & 27 141 & 9 \\
          & & Nematoda & 24 564 & 23 \\
          & & Cnidaria & 14 878 & 36 \\
          & & Porifera & 11 737 & 26 \\
          & & Echinodermata & 10 654 & 14 \\
          & & Bryozoa & 8 631 & 11 \\
          & & Rotifera & 3 093 & 7 \\
          & & Nemertea & 1 793 & 7 \\
          & & Tardigrada & 1 654 & 7 \\
          & & Acanthocephala & 1 596 & 6 \\
          & & Brachiopoda & 1 055 & 9 \\
          & & Nematomorpha & 633 & 7 \\
          & & Chaetognatha & 360 & 7 \\
          & & Hemichordata & 196 & 5 \\ 
          & & Cycliophora & 11 & 5 \\ 
          % 1465207-(1170539+106650+80022+27141+24564+14878+11737+10654+8631+3093+1793+1654+1596+1055+633+360+196+11) = 0          \hline
          Fungi & 254 871 & Ascomycota & 157 045 & 19 \\ 
          & & Basidiomycota & 92 931 & 18 \\
          & & Microsporidia & 1 949 & 6 \\ 
          & & Glomeromycota & 1 490 & 6 \\
          & & Chytridiomycota & 1 456 & 6 \\
          % 254871-157045-92931-1949-1490-1456 = 0
          \hline
          Chloroplastida & 121 239 & Streptophyta & 120 731 & 49 \\
          & & Chlorophyta & 508 & 6 \\
          % 121239-120731-508 = 0
          \hline  
        \end{tabular}
        \caption{\todo{...}}
        \label{table:...} 
      \end{center}  
    \end{table}

  %---------------------------------------------------------------------------------------------------
  %---------------------------------------------------------------------------------------------------
  \section{Missing internal node modeling - Residual table} %\label{sec:Residuals unknown information}
    \begin{table}[h!]
      \begin{center}
          \begin{tabular}{ |l|r|r|r|r| }
            \hline
            \bfseries Model / Taxa & \bfseries Kingdom & \bfseries Phylum & \bfseries Class & \bfseries Order \\
            \hline \hline
            multifurc $\sim$ taxa & 7774454 & \cellcolor{green!10}7435700 & \cellcolor{green!15}7337241 & \cellcolor{green!30}7076068 \\
            \hline
            multifurc $\sim$ taxa + depth & 7752303 & \cellcolor{green!10}7431609 & \cellcolor{green!15}7334754 & \cellcolor{green!30}7027578 \\
            multifurc $\sim$ taxa + max.height & 7730196 & \cellcolor{green!15}7375889 & \cellcolor{green!20}7275856 & \cellcolor{green!30}7005424 \\
            multifurc $\sim$ taxa + min.height & \cellcolor{green!10}7472500 & \cellcolor{green!20}7233486 & \cellcolor{green!25}7144686 & \cellcolor{green!40}6890703 \\
            multifurc $\sim$ taxa + mean.height & \cellcolor{green!15}7304402 & \cellcolor{green!25}7128318 & \cellcolor{green!30}7055313 & \cellcolor{green!40}6815271 \\
            \hline
            multifurc $\sim$ taxa * depth & 7714881 & \cellcolor{green!15}7335396 & \cellcolor{green!20}7250759 & \cellcolor{green!40}6843004 \\
            multifurc $\sim$ taxa * max.height & \cellcolor{green!5}7692980 & \cellcolor{green!15}7311241 & \cellcolor{green!25}7187504 & \cellcolor{green!45}6795823 \\
            multifurc $\sim$ taxa * min.height & \cellcolor{green!10}7442387 & \cellcolor{green!25}7177002 & \cellcolor{green!30}7094933 & \cellcolor{green!45}6795099 \\
            multifurc $\sim$ taxa * mean.height & \cellcolor{green!20}7247309 & \cellcolor{green!30}7020258 & \cellcolor{green!35}6965794 & \cellcolor{green!50}6665565 \\
            \hline
          \end{tabular}
      \end{center}
      \caption{Residuals of multifurcation models \\
        These models were created with the R function \textit{glm()} and compared with the 
          \textit{anova()} function. This results in the listed residuals.}
      \label{table:Residuals multifurcation} 
    \end{table}

  %---------------------------------------------------------------------------------------------------
  %---------------------------------------------------------------------------------------------------
  \section{Missing leaf state modeling - Residual table}\label{sec:Residuals unknown information}
    \begin{table}[h!]
      \begin{center}
        \begin{tabular}{ |l|r|r|r| }
          \hline
          \bfseries Model / Taxa & \bfseries Kingdom & \bfseries Phylum & \bfseries Class \\
          \hline \hline
          multifurc $\sim$ taxa & 545740 & \cellcolor{green!10}499227 & \cellcolor{green!30}482265 \\
          \hline
          multifurc $\sim$ taxa + depth & 544789 & \cellcolor{green!10}493017 & \cellcolor{green!50}478998 \\
          \hline
          multifurc $\sim$ taxa * depth & 544062 & \cellcolor{green!30}488366 & \cellcolor{green!50}476382 \\
          \hline
        \end{tabular}
      \end{center}
      \caption{Residuals of unknown information models \\
        These models were created with the R function \textit{glm()} and compared with the 
          \textit{anova()} function. This results in the listed residuals.}
    \end{table}

  %---------------------------------------------------------------------------------------------------
  %---------------------------------------------------------------------------------------------------
  \section{Cross-validation prediction - Residual table}
    \begin{table}[h]
      \begin{center}
        \begin{tabular}{ |l|r|r|r| }
          \hline
          \bfseries Model / Taxa &\bfseries  Kingdom & \bfseries Phylum & \bfseries Class \\% & Order \\
          \hline \hline
          correctly predicted $\sim$ taxa & 117877 & \cellcolor{green!30}111466 & \cellcolor{green!45}108878 \\% & XXXXX \\
          \hline
          correctly predicted $\sim$ taxa + depth & 117703 & \cellcolor{green!35}110513 & \cellcolor{green!45}108403 \\% & XXXXX \\
          \hline
          correctly predicted $\sim$ taxa * depth & 117592 & \cellcolor{green!40}109827 & \cellcolor{green!50}107994 \\% & XXXXX \\
          \hline
        \end{tabular} 
      \end{center}
      \caption{Residuals of cross-validation prediction \\
        These models were created with the R function \textit{glm()} and compared with the 
          \textit{anova()} function. This results in the listed residuals.}
      \label{table:Residuals cross-validation} 
    \end{table}


\end{document}